### PR TITLE
Timeouts

### PR DESCRIPTION
**Pull request by rickle-msft**
Absolute operation timeouts for requests not expecting a response body. Timeouts based on whether a minimum throughput is maintained throughout the request for requests expecting a response body.

### CODE REVIEW

**Comment by seguler** [comment]
Have you tested this ?

**Comment by rickle-msft** [comment]
My how the tables have turned!! Using the proxy, I tested the absolute timeouts. There's not a good way for me to let data pass through at a specific rate to test the data-rate timeout, but since it's just a curl option that I implemented following their sample and it's pretty straightforward, I trusted that would work very similarly.

File: azure-storage-cpp-lite/include/http/libcurl_http_client.h [hunk 1]
```diff
@@ -148,6 +148,16 @@ namespace microsoft_azure {
                 return m_input_stream;
             }
 
+            void set_absolute_timeout() override {
+                check_code(curl_easy_setopt(m_curl, CURLOPT_TIMEOUT, 30L)); // Timeout of 30 seconds
```
Comment (seguler) [inline azure-storage-cpp-lite/include/http/libcurl_http_client.h]: I think for operations like get_blob_properties we should set this to 5 second. 30 second for those atomic operations is too long

Comment (asorrin-msft) [reply azure-storage-cpp-lite/include/http/libcurl_http_client.h]: I don't know, 30 seconds is fairly standard for us.  It shouldn't be common, but if we get unlucky and hit a partition move, more than 5 seconds can happen

Comment (seguler) [reply azure-storage-cpp-lite/include/http/libcurl_http_client.h]: ok - if it is standard across other things, let's keep it

Comment (seguler) [reply azure-storage-cpp-lite/include/http/libcurl_http_client.h]: Well okay - isn't it actually faster to time out at 5-10 seconds, and then let it retry ? :)

**Pull request merged**